findbugs/src/java/edu/umd/cs/findbugs/ba/vbc/ValueBasedClassIdentifier.java

```diff
@@ -0,0 +1,126 @@
+        JDK_VALUE_BASED_CLASSES.add("java/time/LocalDate");
+        JDK_VALUE_BASED_CLASSES.add("java/time/LocalDateTime");
+        JDK_VALUE_BASED_CLASSES.add("java/time/LocalTime");
+        JDK_VALUE_BASED_CLASSES.add("java/time/MonthDay");
+        JDK_VALUE_BASED_CLASSES.add("java/time/OffsetDateTime");
+        JDK_VALUE_BASED_CLASSES.add("java/time/OffsetTime");
+        JDK_VALUE_BASED_CLASSES.add("java/time/Period");
+        JDK_VALUE_BASED_CLASSES.add("java/time/Year");
+        JDK_VALUE_BASED_CLASSES.add("java/time/YearMonth");
+        JDK_VALUE_BASED_CLASSES.add("java/time/ZonedDateTime");
+        JDK_VALUE_BASED_CLASSES.add("java/time/ZoneId");
+        JDK_VALUE_BASED_CLASSES.add("java/time/ZoneOffset");
+
+        JDK_VALUE_BASED_CLASSES.add("java/time/chrono/HijrahDate");
+        JDK_VALUE_BASED_CLASSES.add("java/time/chrono/JapaneseDate");
+        JDK_VALUE_BASED_CLASSES.add("java/time/chrono/MinguaDate");
+        JDK_VALUE_BASED_CLASSES.add("java/time/chrono/ThaiBuddhistDate");
+
+        /* OTHER */
+        OTHER_VALUE_BASED_CLASSES.add("com/google/common/base/Optional");
+    }
+
+    @StaticConstant
+    private static final ClassDescriptor VALUE_BASED_ANNOTATION_DESCRIPTOR =
+    DescriptorFactory.createClassDescriptor(ValueBased.class);
+
+    /**
+     * Indicates whether the specified class is <i>value-based</i>.
+     * <p>
+     * A class is considered value-based for one of three reasons:
+     * <ul>
+     * <li>the class is part of the JDK and its official documentation identifies it as value-based (e.g.
+     * {@link java.util.Optional Optional})
+     * <li>the class is part of some other widely used library and can be considered as value-based (e.g. <a
+     * href="http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Optional.html">Guava's Optional</a>)
+     * <li>it is annotated with the FindBugs annotation {@link ValueBased @ValueBased}
+     * </ul>
+     *
+     * @param className
+     *            the fully qualified class name in "slashed form" without descriptor and semicolon, e.g. "java/util/Optional"
+     * @return
+     */
+    public static boolean isValueBasedClass(@SlashedClassName String className) {
+        return isJdkValueBasedClass(className) || isOtherValueBasedClass(className) || isAnnotatedAsValueBasedClass(className);
+    }
+
+    private static boolean isJdkValueBasedClass(@SlashedClassName String className) {
+        return JDK_VALUE_BASED_CLASSES.contains(className);
+    }
+
+    private static boolean isOtherValueBasedClass(@SlashedClassName String className) {
+        return OTHER_VALUE_BASED_CLASSES.contains(className);
+    }
+
+    private static boolean isAnnotatedAsValueBasedClass(@SlashedClassName String className) {
+        try {
+            ClassDescriptor classDescriptor = DescriptorFactory.createClassDescriptor(className);
+            XClass xClass = Global.getAnalysisCache().getClassAnalysis(XClass.class, classDescriptor);
```

Comment:
(TODO reminder for myself) Have to check whether this slows down the analysis. This method is executed quite often, but 99.99% of classes are not annotated as VBC. It's possible that non-cached XClass objects are often requested leading to excessive class reloading.